### PR TITLE
Fixed #24067 -- Renamed content types upon model renaming.

### PR DESCRIPTION
**Pull request by charettes**
Thanks to Tim for the extensive review.

### CODE REVIEW

**Comment by charettes** [comment]
This was already mostly reviewed in #5835, the only remaining discussion was about what to do in case of a conflict during rename.

In this case I think the safe approach is simply to silently handle the integrity error and let the `update_contenttypes`'s post_migrate handler prompt the user about what to do next.

File: django/contrib/contenttypes/management.py [hunk 1]
```diff
@@ -1,9 +1,92 @@
...
+            pass
+        else:
+            content_type.model = new_model
+            try:
+                with transaction.atomic(using=db):
+                    content_type.save(update_fields={'model'})
+            except IntegrityError:
+                # Gracefully fallback if a stale content type causes a
+                # conflict as update_contenttypes will take care of the
+                # prompting the user about what should be done next.
```
Comment (timgraham) [inline django/contrib/contenttypes/management.py]: slightly simpler could be: "asking the user what ..."

File: django/contrib/contenttypes/management.py [hunk 2]
```diff
@@ -1,9 +1,92 @@
...
+                content_type.model = old_model
+            else:
+                # Make sure to clear the cache as the `get_by_natual_key()`
```
Comment (timgraham) [inline django/contrib/contenttypes/management.py]: could chop "Make sure to" without any loss of meaning.

File: django/contrib/contenttypes/management.py [hunk 3]
```diff
@@ -1,9 +1,92 @@
...
+    def rename_forward(self, apps, schema_editor):
+        self._rename(apps, schema_editor, self.old_model, self.new_model)
+
+    def rename_backward(self, apps, schema_editor):
+        self._rename(apps, schema_editor, self.new_model, self.old_model)
+
+
+def inject_rename_contenttypes_operations(plan=None, apps=global_apps, using=DEFAULT_DB_ALIAS, **kwargs):
+    """
+    Inserts a `RenameContentType` operation after every planned
```
Comment (timgraham) [inline django/contrib/contenttypes/management.py]: Insert

File: tests/contenttypes_tests/tests.py [hunk 4]
```diff
@@ -445,3 +449,73 @@ def test_multidb(self):
         with self.assertNumQueries(0, using='default'), \
                 self.assertNumQueries(1, using='other'):
             ContentType.objects.get_for_model(Author)
+
+
+@override_settings(
```
Comment (timgraham) [inline tests/contenttypes_tests/tests.py]: Can you use `@modify_settings` instead?

File: tests/contenttypes_tests/tests.py [hunk 5]
```diff
@@ -445,3 +449,73 @@ def test_multidb(self):
...
+    MIGRATION_MODULES=dict(settings.MIGRATION_MODULES, contenttypes_tests='contenttypes_tests.operations_migrations'),
+)
+class ContentTypeOperationsTests(TransactionTestCase):
+    available_apps = ['django.contrib.contenttypes', 'contenttypes_tests']
+
+    def setUp(self):
+        super(ContentTypeOperationsTests, self).setUp()
```
Comment (timgraham) [inline tests/contenttypes_tests/tests.py]: no need for super calls

File: tests/contenttypes_tests/tests.py [hunk 6]
```diff
@@ -445,3 +449,73 @@ def test_multidb(self):
...
+                    self.assertIsInstance(next_operation, contenttypes_management.RenameContentType)
+                    self.assertEqual(next_operation.app_label, migration.app_label)
+                    self.assertEqual(next_operation.old_model, operation.old_name_lower)
+                    self.assertEqual(next_operation.new_model, operation.new_name_lower)
+
+    def test_existing_content_type_rename(self):
+        ContentType.objects.create(app_label='contenttypes_tests', model='foo')
+        stdout = six.StringIO()
+        management.call_command(
+            'migrate', 'contenttypes_tests', database='default', interactive=False, stdout=stdout
```
Comment (timgraham) [inline tests/contenttypes_tests/tests.py]: include trailing commas

File: tests/contenttypes_tests/tests.py (hunk 4 above)
Comment (charettes) [inline tests/contenttypes_tests/tests.py]: Unfortunately I can't as `@modify_settings` only defines action to deal with `list` based settings.

File: tests/contenttypes_tests/tests.py (hunk 6 above)
Comment (timgraham) [inline tests/contenttypes_tests/tests.py]: If you aren't checking the value of `stdout`, can you do `verbosity=0` instead?

Comment (charettes) [inline tests/contenttypes_tests/tests.py]: Will do.

**Pull request merged**